### ISSUE TITLE
Low-CPU usage mode for replay gain scanning

### ISSUE THREAD

**Issue by thejoshwolfe** [andrewrk/groovebasin#194]
If you're running groovebasin on a computer with limited CPU power (or you want to keep groovebasin's CPU usage unobtrusive to the system), there should be a mode that only does just-in-time replay gain scanning. That means when we're already decoding the song in order to play it (or re-encode it for streaming), we use the decoded data to do replay gain scanning, and we adjust the loudness live. This way we get replay gain without any extra decoding CPU usage.

**Comment by alphapapa** [comment]
Here are a couple of suggestions:

1.  Nice down the replaygain threads.  They shouldn't be at priority 0.  They should be no higher than 5, perhaps lower.

2.  A handy mode would be a look-ahead-by-one mode.  For example, I just started using GB on an old netbook, and when I added a bunch of tracks to the playlist, it pegged both cores for quite a while, which spun the fans up loudly for a long time.  It would be nice if there were a mode for GB to scan just the next file in the playlist.  This way it would have the gain ready when the next track starts, but it wouldn't have to scan all the files at once.

**Comment by thejoshwolfe** [comment]
I'm still a fan of doing 1-pass decoding like I proposed in the OP. Like @RazZziel said, slowly adjust the volume to fit the latest understanding of the loudness adjustment needed. The album scanning simplifications work for this strategy, and there'd be no reason to nice anything, since it's all being done by a single thread as slowly as possible.

One drawback to this proposal is that if you don't finish playing through a song, then we interrupt loudness scanning, and have to start over the next time we start playing the song. Also, if we start in the middle of a song, we can't trust that we got a good measurement of the loudness, so we'll not record our findings. Really doing any kind of seek while playing a song makes it so we can't trust out measurements. Only playing from start to finish uninterrupted (pausing is ok) lets us write down the scanning so we don't have to do it again. I'm not too upset about this drawback since the common case is to play a song from start to finish without seeking, and starting the scanning over every time you play a song is not too big of a deal.

Another drawback is the obvious consequences of only looking a short distance ahead: you'll be able to hear the song adjusting its volume as you get further along in the song. We can't see the future, so we'll inevitably guess wrong sometimes. This will be especially noticeable in any song with a long crescendo intro like [Kill the Noise Part 1 by Kill the Noise](https://www.youtube.com/watch?v=cyiB_rkob8o) or [Intro by Muse from Absolution](https://www.youtube.com/watch?v=rdfgxrYRFiw). I think if we do gradual volume adjustments, this issue should be pretty forgivable. I used MPD's volume adjustment that followed this algorithm, and while I noticed it and didn't like it, it was still pretty decent.

One more feature might be nice, which is making some attempt at predicting the song length without having to actually scan the entire thing. Often, songs get imported with a duration of negative 5 million hours or something ridiculous, and Groove Basin just displays that until we scan the whole thing. Negative song durations make the seek bar not work at all (cursor stays at 0 the whole song, and seeking always goes to t=0.). A feature that my change should probably do is try to predict the song length while we're scanning it (probably extrapolate the time/bytes rate with the total file size.). This paragraph could be a separate github issue.

**Comment by thejoshwolfe** [comment]
To clarify, I'm proposing the above for the "Low CPU Mode", which is a different mode from normal. We should still do the eager scanning mode sometimes, possibly use a heuristic to choose which mode to be in like @andrewrk is suggesting, or give the user a choice, or both.

Here is my proposal for mode choice/switching. Give admins these choices in the Settings pane:

When should Groove Basin scan future songs?

1. Never. Only scan songs while playing them. (This is my Low CPU Mode proposed above.)
2. While Groove Basin is paused. (This is good for dedicated but weak machines.)
3. All the time. (This is status quo.)

What future songs should Groove Basin scan?

a. Only the next song. (proposed by @alphapapa above.)
b. All the songs in the current queue. (This is status quo.)
c. The songs in the current queue, and then all the songs in the library.

The default initial settings will be 3 and b (status quo) for machines that have more than 1 CPU core. Otherwise, the defaults will be 2 and b.

This proposal introduces a few new ideas that could be their own github issues.

**Comment by alphapapa** [comment]
Well, I'm not the one doing the work, so it doesn't matter what I think.  :)  But here is my two cents:

@thejoshwolfe has put a lot of thought into those ideas, but I think it is not the best plan.

1.  Scanning the currently playing track is not a good idea.  The default behavior should be to have loudness data ready before a track begins playing.  The whole point of ReplayGain is that the loudness is correct for the whole track so it doesn't seem to get too loud or too soft.  But if you defer scanning until the track begins playing, the track will play at the wrong loudness until the scanning finishes--and that defeats the whole purpose of ReplayGain.  Also, scanning the track can take a while on slower systems, so that could cause a track to play at the wrong loudness for quite some time.

The exception to this should be that, if a track begins playing, and it lacks loudness data, it should be scanned as the track plays.  But this should be a last resort.  It should generally only happen if a user adds tracks to the library and immediately starts playing them, before GB has a chance to scan them.

2.  The scanning threads definitely *should* be niced down by default.  I do not ever want my CPU bogged down by threads doing work in the background.  This is especially important for low-powered systems that are supposed to be silent.

3.  I'm not sure about the proposal to use multiple threads on any system with more than 1 CPU.  For example, I have two relatively low-powered netbooks, one of which I use as a jukebox.  They both have dual-core CPUs.  They are slow, but they are dual-core.  If the system is otherwise idle, it would be okay to have two scanning threads, but if I'm using the system, that might be an issue.  (Although if they are niced down properly, it should be okay.)

4.  I think the simplest, most sensible default is to scan one track ahead.  This will help avoid running systems too hot and loud which are supposed to remain cool and quiet (like ones whose primary job is to play music).

On top of that, there could be two options:

a) Scan entire queue.  This would ensure that all the tracks in the queue are scanned ASAP, anytime the queue changes.  This could also be a button/command in the UI/protocol to do it on-demand.

b) Scan entire library.  This would ensure that all the tracks in the library are scanned ASAP, anytime new tracks are detected.  And this could also be a button/command in the UI/protocol to do it on-demand.

5.  I would suggest avoiding any kind of modal behavior for this issue.  That makes it more complicated, trying to figure out and remember which settings the mode affects and overrides, etc.  With my proposal, there are only 2 settings:

* Scanning thread niceness.  Should probably default to 5 or 10.
* Number of tracks ahead to scan.  This would have 3 possible choices: a number, `queue`, and `library`.  The number could range from 0-anything, according to the user's desires.  If set to 0, tracks are not scanned until they begin playing; most users probably won't choose this setting.  If set to 1 or more, GB scans tracks that many positions ahead in the queue.  If the user sets it to `queue`, the the whole queue is kept scanned.  If he sets it to `library`, the entire library is kept scanned.
 
So, in conclusion: cool, quiet, and low-priority scanning by default, with optional immediate and constant queue and/or library scanning.

**Comment by thejoshwolfe** [comment]
> But if you defer scanning until the track begins playing, the track will play at the wrong loudness until the scanning finishes

That was not my proposal. The scan-as-you-go method I was proposing would constantly adjust the loudness of the song as the analysis is performed. I think in practice this would result in asymptotically approaching the appropriate loudness as the song progresses, often getting indistinguishably close in the first few seconds. It's also the optimal amount of effort, because you'd only read the file from the file system once, decode the audio once, and then use the decoded data for both playback and analysis.

I think I agree with your points about nicing down scanning threads. Seems like an appropriate usecase for nice.

The rest of your proposal seems pretty compatible with mine.

**Comment by alphapapa** [comment]
> That was not my proposal. The scan-as-you-go method I was proposing would constantly adjust the loudness of the song as the analysis is performed. I think in practice this would result in asymptotically approaching the appropriate loudness as the song progresses, often getting indistinguishably close in the first few seconds

Sorry, I don't think so.  Think of a piece of classical music that starts off quietly for the first minute or two.  Until the scanning had progressed past that point, it would appear that the track was very quiet, and the volume would be increased too much to compensate.  That would cause the initial part of the track to play far too loudly.  The same could happen in reverse with a track that starts off loudly and gets quieter.  And some tracks simply have a wide dynamic range with loud and quiet parts throughout.  Classical orchestra tracks often have very wide dynamic ranges.

I just don't think that method will work well in practice.  It's impossible to know the correct gain until the track has been completely scanned, so that would, by definition, cause the track to play at the wrong loudness until the scanning is finished.

If you want to do a little experiment, find a long classical track with a wide dynamic range.  Divide it into overlapping segments of increasing length (e.g. segment 1=0-1 minute, segment 2=0-2 minutes, etc).  Save the segments as separate files and run replaygain on each file.  Then divide each segment so they no longer overlap (i.e. delete the 0-1 minute portion of segment 2).  Make sure to keep the replaygain of each segment.  Then play each segment in a playlist with in track mode (not album mode).

What I think you will find is that each segment plays at a very different loudness setting due to each segment having been analyzed independently, rather than the entire track being analyzed at once.  This will roughly simulate the effects of your proposed method.

For some genres and some tracks, the effect might not be noticeable.  But for many genres and tracks, I think it would be very disruptive to the listening experience.

**Comment by diogocp** [comment]
I agree with @alphapapa. Think about what what Ravel's Boléro would sound like using that algorithm. It begins pianissimo, so the gain would be high. Then it slowly builds up to a  fortissimo, so you would gradualy reduce the gain. Instead of a long crescendo, the loudness would be pretty flat throughout.
